### PR TITLE
[Submission] Executable tutorial: How to setup a CI/CD pipeline for a react application

### PR DESCRIPTION
**Pull request by damnePers**
Submission of Executable Tutorial.

Proposal PR #1263 
Feedback PR #1267 

## Memeber

Sara Damne, damne@kth.se, [@damnePers](https://github.com/damnePers)

## Proposal

This tutorial will teach the user to configure a CI/CD pipeline for a react app using Heroku and CircleCI. 

The user will learn how to:
- Setup react app
- Configure Heroku
- Configure CircleCI
- Configure Heroku

The tutorial will be published as a medium article.
Note that this is different from the proposal. 

## Links

Medium article 
- https://medium.com/saradamne/how-to-setup-a-ci-cd-pipeline-for-a-react-application-cfdc511e893f

## Grading Cirteria

|                                             | Yes | No | Remarkable |
|-------------------------------------------- | ----|----|-------------|
|The TA can successful execute all the commands of the tutorial (mandatory) | Yes :heavy_check_mark: | No | In the browser  |
|If local execution, runs on Linux | Yes :heavy_check_mark: | No | Easy to setup and run :heavy_check_mark:  |
|The tutorial gives enough background | Yes :heavy_check_mark: | No | Comprehensive background :heavy_check_mark: |
|The tutorial is easy to follow  | Yes :heavy_check_mark: | No | Well documented :heavy_check_mark: |
|The tutorial is original, no such tutorial exists on the web | Yes | No | The teaching team never heard about it |
|The tutorial contains [easter eggs](https://github.com/OrkoHunter/python-easter-eggs) | Yes | No | Subtle and fun |
|The tutorial is successful (attracts comments and success) | Yes | No | Lively discussion |
|The language is correct | Yes :heavy_check_mark:| No | Interesting narrative :heavy_check_mark: |

### CODE REVIEW

**Comment by Grunler** [comment]
Hello, we are currently working on the feedback but we're experiencing some problems in step 3 as the tab crashes on the circleCI projects page.

We will try again later but I just wanted to give you a heads up that we might not be able to provide feedback for every part of the tutorial unless the issue is resolved soon.

**Comment by damnePers** [comment]
Hi @Grunler, interesting... I realized that I hadn't linked to CircleCI so that should be updated at least.
What do you mean that the tab crashes? Is it when you have pushed a commit and it is running or do you not get access to the project page at all?

**Comment by LeeBadal** [comment]
@damnePers CircleCI was having some issues at the time, they should be resolved now. As of this I will post my feedback now in a comment below and @Grunler will post his feedback later.

**Comment by LeeBadal** [comment]
# Feedback for "How to setup a CI/CD pipeline for a react application", part 1:
Link to Medium post: https://medium.com/saradamne/how-to-setup-a-ci-cd-pipeline-for-a-react-application-cfdc511e893f
## Feedback Members
Lee Badal (badal@kth.se)

GitHub: [@LeeBadal](https://github.com/LeeBadal)
##### Message:
Good job! I enjoyed the read and managed to successfully create a CI/CD pipeline. (which was the goal!)

Here is my feedback, I like to provide it in full sentences so my thought process can be understood, if anything needs to be clarified let me know. Happy to have a discussion about things.

## Feedback from Lee

The topic is interesting from the perspective of an individual that enjoys making small projects and really wants to jumpstart CI/CD. The choice of tools is good since they are easy to use at a basic level and the amount of downloads to get it all running wasn't overwhelming. The goal and the tools are clearly explained, I found the introduction to CI/CD slightly dull, it did not get me excited about the practice even though it was instructive.

The post was visually pleasing, with a nice split between text and images. However, I would have preferred being able to copy the code for the config.yml file, tabs instead of space made it break which led me to have to make changes and push the changes. I would therefore just post this as a code block which you have done previously.

At some point in the tutorial I stumbled upon this line:
***"Go ahead and create a Github repo and push your changes."***
Which in my head didn't make sense, this is perhaps my personal workflow of first creating a clean repo then adding files. Otherwise, maybe you could add a short comment about how in an easy way create a repo from this state (if you didn't use GitHub UI).

From a React developer perspective, perhaps the first steps are a little too trivial. And for a non-react developer or a beginner, perhaps a CI/CD pipeline is not the most important task to focus on. Therefore I do not really know who the target audience is(besides TA's / feedback givers). Even though I found it is an interesting read, perhaps it could be clarified who the tutorial is for.

It would have been great with an image or clear description of the folder structure.
***"First add a directory .circleci to the root of your name-of-your-project folder."***
Was a sentence that was unnecessarily hard to understand.

In ***"Step 3 — Connect to CirleCI"*** it would have been nice with a link to circleCI, this is the first time you have to visit to complete the tutorial and you should not have to scroll to find the link. 
 
I completed the tutorial after 34 minutes, that being said I already had a Heroku account & npm installed. I did make some mistakes and CircleCI was having a few issues at the time. I can see how you can speed run the tutorial if you want to, but 10 minutes was slightly optimistic. I don't think it is necessary to mention the time it should take.

## TLDR suggested changes;
* Convert image of code to block of code (to be able to copy)
* Add a link to CircleCI in step 3
* Reestimate/remove estimated time being 10 minutes.
## Suggested but debatable changes:
* Narrow down to a specific target audience
* Use less formal titles/language
* Explain why these tools are good/bad and give a more wholesome picture of the task at hand
* Review where creating a GitHub repo should be done.

**Comment by Grunler** [comment]
# Feedback for "How to setup a CI/CD pipeline for a react application", part 2:

## Feedback member:

Daniel Grünler (grunler@kth.se)

GitHub: [@Grunler](https://github.com/Grunler)

## Overall impression:

I enjoyed reading and executing this tutorial. I had not used Heroku or CircleCI before but apart from some unfortunate error on the CircleCI side I had no issue following the tutorial and I enjoyed learning something new.

The executable tutorial is a well-written and streamlined take on an interesting and useful subject. By focusing on a simple case and assuming that the reader is already familiar with using the command line and Github you avoid getting side tracked. This makes for a short tutorial that is easy to follow given those assumptions. I can imagine that the tutorial would be hard to follow for those who are not used to using Github and in particular initializing a local repo and pushing it to Github. Overall I think your approach was correct but you could add some additional links and clarifications in some places. Additionally there are some minor things that you could fix and improve upon. I’ll comment on these in more detail below.


### Time estimate and Overview

The estimated time seems optimistic. The tutorial is short but due to build times and some account setup I would estimate it at 20 minutes at a minimum.

The overview is short and to the point. I would not remove or add anything to it. The only minor suggestions I have is to capitalize the R in React and to fix the typo on Heroku.

### Introduction to CI/CD

The introduction to CI/CD also looks good to me. There is no need for a more thorough explanation in an executable tutorial like this and your explanation is sufficient for those with minimal knowledge of the terms to get an idea of what they mean. 


### Prerequisites

I would add “access to a UNIX computer” or similar to the list of prerequisites as the tutorial requires a UNIX terminal to follow. I appreciate that you included links to npm and Heroku to make the setup quicker. Finally I would suggest changing the title from “Prerequisite” to “Prerequisites”.

### Step 1 — Set up a react application

The instructions in this section were very simple to follow. The most complex part would be pushing the local files to a Github repo and it is not explained at all in the section. I understand why you would want to keep it outside the scope of the tutorial so as to not get sidetracked, but you could always link a resource such as this one:
https://kbroman.org/github_tutorial/pages/init.html

You could also consider changing the tutorial slightly by asking the reader to create the Github project themselves and cloning it locally at the start of the first step.


### Step 2: 

I would suggest that you include the contents of the .yml file as text as a compliment to the photo. There are multiple good reasons for doing so. The photo might be down or not able to be displayed for some reason and it makes it easier to copy the contents and it makes it easier for sight impaired readers to follow the tutorial. Another option is to add alternative text to the image.

Here are a couple useful links:
https://www.w3.org/WAI/tutorials/images/textual/#image-of-styled-text-with-decorative-effect
https://medium.com/gobeyond-ai/adding-alt-tags-to-images-on-medium-563d6d31130a

You could consider adding alternative text to all images but this one is the most important in my opinion.


### Step 3: Connect to CircleC
This section is great, clear and with a lot of images. I originally had some issues such as one image missing and there not being a link to the CircleCI website (the link was in step two when it was not needed). Both these issues have already been fixed while I’m writing this feedback!

### Step 4: Connect Heroku and deploy app

This section was also very clear and it was easy to follow.

### Conclusion

Your final remarks for the tutorial were suitable as you summed it up and encouraged the reader to proceed to take more steps on their journey to automation.

**Comment by damnePers** [comment]
Thanks for the very good feedback! 

Here are the changes I've made

- Removed estimated time
- Fixed typos in the overview
- Switched out image of code to code snippet
- Added a folder structure overview in step 2
- Add link to CircleCI in step 3 and removed this link from step 2
- Added more prerequisites - git/github skills, unix terminal
- Added captures to images
- Changed the order of when to create a Github repo for the project.
- Added links to tutorials/documentation for users with less experience
- Added step-overview to make it easier for users with more experience skip ahead
- Changed some of the step-names to make them less formal
- Added a sentence about why this tutorial is useful in the overview
